### PR TITLE
Add a new semodule back end that forks out to policycoreutils to avoid memory leaks

### PR DESCRIPTION
**Pull request by jhrozek**
During stress-testing of SPO it was found that selinuxd leaks memory. It
appears that the leak is coming from libsemanage (confirmed with
valgrind and fixed upstream). Using the policycoreutils back end,
testing on OpenShift 4.x, the selinuxd containers spike at around ~400MB
before falling back to ~30-50 MB. With the semanage back end, the memory
would climb steadily until it would hit the 1GB limit imposed by SPO's
DS definition and be OOMKilled.

To work around that, this PR adds a new back-end for the semodule
package that shells out to the policycoreutils utilities. This back end
is the default on "RHEL-like" distributions, but otherwise selectable
during build using the SEMODULE_BACKEND environment variable that
translates into a build tag.

The "Handler" interface was moved from semodule top-level to
semodule/seiface to avoid circular dependencies.

### CODE REVIEW

**Comment by jhrozek** [comment]
Local e2e tests with vagrant succeeded, submitting to make sure we pass CI. But please don't merge this yet even if the PR passes because:
 1) The SPO policies must be updated first
 2) I'd like to make sure the policy updates are minimal and the selinux policy additions in the capabilities are giving me the creeps.

**Comment by jhrozek** [comment]
> Local e2e tests with vagrant succeeded, submitting to make sure we pass CI. But please don't merge this yet even if the PR passes because:
> 
>     1. The SPO policies must be updated first

OK, this apparently does not need to be done.

**Comment by jhrozek** [comment]
Great, CI passes, so this is a 'known good' state. I'll try trimming the policy a little bit to see if there are any excessive syscalls allowed and remove the DNM tag when I'm done.

**Comment by jhrozek** [comment]
OK, I removed all syscalls that seemed spurious from the seccomp profile, including `keyctl`, `mount`, `pivor_root` and `unshare`. The ones that are added seem harmless (`fchown` is an exception to this but it seems to be required by the policycoreutils back end and `setuid` seems to be required by the container engine..)

**Comment by JAORMX** [comment]
> OK, I removed all syscalls that seemed spurious from the seccomp profile, including `keyctl`, `mount`, `pivor_root` and `unshare`. The ones that are added seem harmless (`fchown` is an exception to this but it seems to be required by the policycoreutils back end and `setuid` seems to be required by the container engine..)

How did you verify that this was fine? A lot of these were generated via the SPO, so they might have been needed by crun/runc at the time. Would be good to add an SPO e2e test here so we verify we're not breaking anything there.

File: Makefile
Comment (JAORMX) [inline Makefile:31]: thought you wanted to default to `policycoreutils` already

File: pkg/semodule/seiface/interface.go
Comment (JAORMX) [inline pkg/semodule/seiface/interface.go:1]: nit: this module name seems quite redundant. Why not call this just `interface` instead? We know it's related to `semodule` since that's the parent package name anyway.

File: pkg/semodule/policycoreutils/policycoreutils.go
Comment (JAORMX) [inline pkg/semodule/policycoreutils/policycoreutils.go:22]: Could you use `CommandContext` instead? This way we can pass a `context.Context` object and potentially do timeouts. It could be `context.TODO` to begin with and we can leave for later (a low hanging fruit bug would be ideal for newcomers to the project).

File: pkg/semodule/policycoreutils/policycoreutils.go
Comment (JAORMX) [inline pkg/semodule/policycoreutils/policycoreutils.go:38]: doesn't the command that was run already appear in the `err` that was returned? If so, might want to change the error message to something more human readable that we can `grep` for.

File: pkg/semodule/policycoreutils/policycoreutils.go
Comment (JAORMX) [inline pkg/semodule/policycoreutils/policycoreutils.go:42]: same human readability comment as above

File: pkg/semodule/policycoreutils/policycoreutils.go
Comment (JAORMX) [inline pkg/semodule/policycoreutils/policycoreutils.go:47]: why not use the `runSeModule` function you define above?

File: pkg/semodule/policycoreutils/policycoreutils.go
Comment (JAORMX) [inline pkg/semodule/policycoreutils/policycoreutils.go:21]: you could make this more general by using variable arguments. e.g.

```
func runSeModule(args ...string) (string, error) {
```

File: pkg/semodule/policycoreutils/policycoreutils.go
Comment (JAORMX) [reply pkg/semodule/policycoreutils/policycoreutils.go:22]: Also... should you make this call verbose so we can get more info on potential issues?

**Review by JAORMX — CHANGES_REQUESTED** [review]
If we're gonna keep both backends for a while, I'd like to see e2e tests for policycoreutils too. Unit tests would be ideal as well, but I understand that might be challenging.

**Comment by jhrozek** [comment]
> > OK, I removed all syscalls that seemed spurious from the seccomp profile, including `keyctl`, `mount`, `pivor_root` and `unshare`. The ones that are added seem harmless (`fchown` is an exception to this but it seems to be required by the policycoreutils back end and `setuid` seems to be required by the container engine..)
> 
> How did you verify that this was fine? A lot of these were generated via the SPO, so they might have been needed by crun/runc at the time. Would be good to add an SPO e2e test here so we verify we're not breaking anything there.

I ran the e2e tests on an OCP cluster. Given that we don't have an OCP e2e test in the SPO suite, are you suggesting we add a CentOS/AlmaLinux/whatever based e2e test for SPO like we do for Fedora?

**Comment by JAORMX** [comment]
> > > OK, I removed all syscalls that seemed spurious from the seccomp profile, including `keyctl`, `mount`, `pivor_root` and `unshare`. The ones that are added seem harmless (`fchown` is an exception to this but it seems to be required by the policycoreutils back end and `setuid` seems to be required by the container engine..)
> > 
> > 
> > How did you verify that this was fine? A lot of these were generated via the SPO, so they might have been needed by crun/runc at the time. Would be good to add an SPO e2e test here so we verify we're not breaking anything there.
> 
> I ran the e2e tests on an OCP cluster. Given that we don't have an OCP e2e test in the SPO suite, are you suggesting we add a CentOS/AlmaLinux/whatever based e2e test for SPO like we do for Fedora?

I was actually suggesting to add a test that runs the SPO on this repo. Note that I do think we should have an AlmaLinux test in the SPO repo too.

File: pkg/semodule/policycoreutils/policycoreutils.go
Comment (jhrozek) [reply pkg/semodule/policycoreutils/policycoreutils.go]: Thanks, that was refactoring gone wrong.

File: pkg/semodule/policycoreutils/policycoreutils.go
Comment (jhrozek) [reply pkg/semodule/policycoreutils/policycoreutils.go]: Done

File: Makefile
Comment (jhrozek) [reply Makefile]: Initially I wanted to only change centos builds, because I think that fedora builds are fixed, but a) I have not verified that outside a local VM and b) it makes sense to keep one default. So, done.

**Comment by jhrozek** [comment]
e2e test failed because of gofmt. Will fix later today.

**Comment by JAORMX** [comment]
Are the tests taking longer now? This sounds like a regression 😕

**Comment by jhrozek** [comment]
> Are the tests taking longer now? This sounds like a regression confused

Yes, the policycoreutils back end is slower somewhat. But that's not surprising, is it.

**Pull request merged**
